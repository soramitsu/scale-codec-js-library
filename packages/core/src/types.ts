export type DecodeResult<T> = [value: T, decodedBytes: number];

export type Decode<T> = (bytes: Uint8Array) => DecodeResult<T>;

export type Encode<T> = (value: T) => Uint8Array;

export interface Codec<T> {
    encode: Encode<T>;
    decode: Decode<T>;
}
