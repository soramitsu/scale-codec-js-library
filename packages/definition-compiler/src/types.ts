export interface RenderNamespaceDefinitionParams {
  /**
   * Runtime library with STD codecs + reexports from the core library.
   *
   * Defaults to `@scale-codec/definition-runtime`
   */
  runtimeLib?: string

  /**
   * Types available in the runtime library.
   */
  runtimeTypes?: Set<string>

  /**
   * Single tuples are always an arrays with a single element. It is possible to make final code cleaner
   * (and a bit performant) if render such tuples just as aliases for the inner element. It is optional feature.
   */
  rollupSingleTuplesIntoAliases?: boolean

  /**
   * @default 'Void'
   */
  typeForVoidAliasing?: string

  /**
   * Enable sorting of types to minimize amount of allocated dynamic references between them
   *
   * @default false
   * @experimental
   */
  optimizeDyns?: boolean
}

/**
 * The main map with all the definitions
 *
 * @remarks
 * Name of type should be **a valid JavaScript identifier name**, because it will be compiled to an identifier.
 *
 * @example
 * ```ts
 * const definition: NamespaceDefinition = {
 *   Person: {
 *     t: 'struct',
 *     fields: [
 *       { name: 'name', ref: 'str' },
 *       { name: 'age', ref: 'u8' }
 *     ]
 *   },
 *   VecPerson: { t: 'vec', item: 'Person' }
 * }
 * ```
 */
export type NamespaceDefinition = Record<string, TypeDef>

export type TypeRef = string

/**
 * Just an alias to the inner type
 */
export type DefAlias = {
  ref: TypeRef
}

/**
 * Fixed-length array
 */
export type DefArray = {
  /**
   * Inner type name
   */
  item: TypeRef
  len: number
}

/**
 * It's like {@link DefArray} but for bytes (u8). Use it for bytes for better performance.
 */
export type DefBytesArray = {
  len: number
}

/**
 * `Vec<T>` definition
 */
export type DefVec = {
  /**
   * Inner vec type name
   */
  item: TypeRef
}

/**
 * Tuple definition
 */
export type DefTuple = {
  /**
   * Array of inner types
   */
  items: TypeRef[]
}

/**
 * Structure definition
 */
export type DefStruct = {
  /**
   * @remarks
   * **note**: order of fields matters!
   */
  fields: DefStructField[]
}

export type DefStructField = {
  /**
   * Name of the struct field
   */
  name: string
  /**
   * Reference to the type
   */
  ref: TypeRef
}

/**
 * Map definition (e.g. `HashMap`, `BTreeMap`)
 */
export type DefMap = {
  key: TypeRef
  value: TypeRef
}

/**
 * Set definition (e.g. `HashSet`, `BTreeSet`)
 */
export type DefSet = {
  entry: TypeRef
}

/**
 * Enum definition
 */
export type DefEnum = {
  /**
   * @remarks
   * Order of variants doesn't matter, but variants should not contain collisions between their names
   * and discriminants
   */
  variants: DefEnumVariant[]
}

export type DefEnumVariant = {
  name: string
  discriminant: number
  /**
   * No ref/null ref means that this variant is empty
   */
  ref?: TypeRef | null
}

/**
 * Option enum definition
 */
export type DefOption = {
  some: TypeRef
}

/**
 * Result enum definition
 */
export type DefResult = {
  ok: TypeRef
  err: TypeRef
}

/**
 * External type definition - import codec from external module
 *
 * @remarks
 * Provides a possibility to define external types, e.g. to use some complex structure from another compiled namespace
 * OR to define your own custom low-level codec for type that is not included into the SCALE codec spec by default.
 *
 * Note that the import should be of `FragmentBuilder` type.
 */
export type DefImport = {
  /**
   * Where to import from, path
   *
   * @example
   * ```ts
   * import { ... } from '<here is the module name>'
   * ```
   */
  module: string
  /**
   * Name of the type inside of the module. If this field is omitted, the own type name will be used
   *
   * @todo *define custom name for each import?*
   */
  nameInModule?: string
}

type WithTMark<T, M extends string> = T & {
  t: M
}

export type TypeDef =
  | WithTMark<DefAlias, 'alias'>
  | WithTMark<DefArray, 'array'>
  | WithTMark<DefBytesArray, 'bytes-array'>
  | WithTMark<DefVec, 'vec'>
  | WithTMark<DefTuple, 'tuple'>
  | WithTMark<DefStruct, 'struct'>
  | WithTMark<DefMap, 'map'>
  | WithTMark<DefSet, 'set'>
  | WithTMark<DefEnum, 'enum'>
  | WithTMark<DefOption, 'option'>
  | WithTMark<DefResult, 'result'>
  | WithTMark<DefImport, 'import'>
