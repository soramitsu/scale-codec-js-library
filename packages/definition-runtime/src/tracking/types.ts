import { Decode, Walker } from '@scale-codec/core'

/**
 * This symbol could be used by tracker to inspect value
 */
export const TrackValueInspect = Symbol('TrackValueInspect')

/**
 * Implement this for the object if you want to handle its inspection in the console
 */
export interface TrackValueInspectable {
  [TrackValueInspect]: () => any
}

export type TrackDecodeFn = <T>(loc: string, walker: Walker, decode: Decode<T>) => T

export type RefineDecodeLocFn = <T>(loc: string, headlessDecode: () => T) => T

/**
 * Instance that defines method to track decoding flow
 */
export interface CodecTracker {
  /**
   * Used to track decode step: location, input, and result
   */
  decode?: TrackDecodeFn
  /**
   * Used to clarify decode location without tracking of the input and the output
   */
  refineDecodeLoc?: RefineDecodeLocFn
}
