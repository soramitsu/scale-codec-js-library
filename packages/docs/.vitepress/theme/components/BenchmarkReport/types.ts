export interface ReportData {
    name: string
    date: string
    version: string | null
    results: Result[]
    fastest: ReportCaseScore
    slowest: ReportCaseScore
}

export interface Result {
    name: string
    ops: number
    margin: number
    percentSlower: number
}

export interface ReportCaseScore {
    name: string
    index: number
}
