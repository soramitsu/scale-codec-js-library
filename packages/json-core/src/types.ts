export type JsonValue = null | string | boolean | number | { [K in string]: JsonValue } | JsonValue[];

export type Serialize<T> = (value: T) => JsonValue;

export type Deserialize<T> = (raw: JsonValue) => T;

export interface Serde<T> {
    ser: Serialize<T>;
    de: Deserialize<T>;
}
