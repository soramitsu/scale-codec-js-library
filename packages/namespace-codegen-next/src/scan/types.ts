export type ScanParams = {
    isKnownType: (ref: string) => boolean;
};
