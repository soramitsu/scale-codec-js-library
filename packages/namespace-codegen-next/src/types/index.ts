export * from './definition';
export * from './heap';
