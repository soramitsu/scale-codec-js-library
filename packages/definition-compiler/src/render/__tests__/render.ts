import { renderNamespaceDefinition } from '..'
import { NamespaceDefinition, RenderNamespaceDefinitionParams } from '../../types'

function renderFactory(def: NamespaceDefinition, params?: RenderNamespaceDefinitionParams): string {
    return renderNamespaceDefinition(def, params)
}

it('Renders vec', () => {
    expect(
        renderFactory({
            Vec_str: {
                t: 'vec',
                item: 'Str',
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { Str, createVecCodec } from '@scale-codec/definition-runtime'

        import type { ArrayCodecAndFactory, Opaque } from '@scale-codec/definition-runtime'

        // Type: Vec_str

        type Vec_str__actual = Str[]

        interface Vec_str extends Opaque<Vec_str__actual, Vec_str> {}

        const Vec_str: ArrayCodecAndFactory<Vec_str__actual, Vec_str> = createVecCodec<Vec_str__actual, Vec_str>('Vec_str', Str)

        // Exports

        export { Vec_str }"
    `)
})

it('Renders struct + tuple', () => {
    expect(
        renderFactory({
            Structural: {
                t: 'struct',
                fields: [
                    {
                        name: 'nums',
                        ref: 'Tuple_u8_i128',
                    },

                    {
                        name: 'mur',
                        ref: 'Bool',
                    },
                ],
            },

            Tuple_u8_i128: {
                t: 'tuple',
                items: ['U8', 'I128'],
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { Bool, I128, U8, createStructCodec, createTupleCodec, dynCodec } from '@scale-codec/definition-runtime'

        import type { ArrayCodecAndFactory, Opaque, StructCodecAndFactory } from '@scale-codec/definition-runtime'

        // Dynamic codecs

        const __dyn_Tuple_u8_i128 = dynCodec(() => Tuple_u8_i128)

        // Type: Structural

        interface Structural__actual {
            nums: Tuple_u8_i128
            mur: Bool
        }

        interface Structural extends Opaque<Structural__actual, Structural> {}

        const Structural: StructCodecAndFactory<Structural__actual, Structural> = createStructCodec<Structural__actual, Structural>('Structural', [
            ['nums', __dyn_Tuple_u8_i128],
            ['mur', Bool]
        ])

        // Type: Tuple_u8_i128

        type Tuple_u8_i128__actual = [U8, I128]

        interface Tuple_u8_i128 extends Opaque<Tuple_u8_i128__actual, Tuple_u8_i128> {}

        const Tuple_u8_i128: ArrayCodecAndFactory<Tuple_u8_i128__actual, Tuple_u8_i128> = createTupleCodec<Tuple_u8_i128__actual, Tuple_u8_i128>('Tuple_u8_i128', [U8, I128])

        // Exports

        export { Structural, Tuple_u8_i128 }"
    `)
})

it('Renders enum', () => {
    expect(
        renderFactory({
            Message: {
                t: 'enum',
                variants: [
                    {
                        name: 'Quit',
                        discriminant: 0,
                    },

                    {
                        name: 'Greeting',
                        discriminant: 1,
                        ref: 'Str',
                    },
                ],
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { Enum, Str, createEnumCodec } from '@scale-codec/definition-runtime'

        import type { EnumCodecAndFactory, Opaque } from '@scale-codec/definition-runtime'

        // Type: Message

        type Message__actual = Enum<
            | 'Quit'
            | ['Greeting', Str]
        >

        interface Message extends Opaque<Message__actual, Message> {}

        const Message: EnumCodecAndFactory<Message> = createEnumCodec<Message__actual, Message>('Message', [
            [0, 'Quit'],
            [1, 'Greeting', Str]
        ])

        // Exports

        export { Message }"
    `)
})

it('Renders set', () => {
    expect(
        renderFactory({
            Set_Message: {
                t: 'set',
                entry: 'Message',
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { createSetCodec, dynCodec } from '@scale-codec/definition-runtime'

        import type { Opaque, SetCodecAndFactory } from '@scale-codec/definition-runtime'

        // Dynamic codecs

        const __dyn_Message = dynCodec(() => Message)

        // Type: Set_Message

        type Set_Message__actual = Set<Message>

        interface Set_Message extends Opaque<Set_Message__actual, Set_Message> {}

        const Set_Message: SetCodecAndFactory<Set_Message__actual, Set_Message> = createSetCodec<Set_Message__actual, Set_Message>('Set_Message', __dyn_Message)

        // Exports

        export { Set_Message }"
    `)
})

it('Renders map', () => {
    expect(
        renderFactory({
            Map_str_Message: {
                t: 'map',
                key: 'Str',
                value: 'Message',
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { Str, createMapCodec, dynCodec } from '@scale-codec/definition-runtime'

        import type { MapCodecAndFactory, Opaque } from '@scale-codec/definition-runtime'

        // Dynamic codecs

        const __dyn_Message = dynCodec(() => Message)

        // Type: Map_str_Message

        type Map_str_Message__actual = Map<Str, Message>

        interface Map_str_Message extends Opaque<Map_str_Message__actual, Map_str_Message> {}

        const Map_str_Message: MapCodecAndFactory<Map_str_Message__actual, Map_str_Message> = createMapCodec<Map_str_Message__actual, Map_str_Message>('Map_str_Message', Str, __dyn_Message)

        // Exports

        export { Map_str_Message }"
    `)
})

it('Renders array', () => {
    expect(
        renderFactory({
            Array_Item_15: {
                t: 'array',
                item: 'Item',
                len: 15,
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { createArrayCodec, dynCodec } from '@scale-codec/definition-runtime'

        import type { ArrayCodecAndFactory, Opaque } from '@scale-codec/definition-runtime'

        // Dynamic codecs

        const __dyn_Item = dynCodec(() => Item)

        // Type: Array_Item_15

        interface Array_Item_15__actual extends Array<Item> {}

        interface Array_Item_15 extends Opaque<Array_Item_15__actual, Array_Item_15> {}

        const Array_Item_15: ArrayCodecAndFactory<Array_Item_15__actual, Array_Item_15> = createArrayCodec<Array_Item_15__actual, Array_Item_15>('Array_Item_15', __dyn_Item, 15)

        // Exports

        export { Array_Item_15 }"
    `)
})

it('Renders bytes array', () => {
    expect(
        renderFactory({
            Hash: {
                t: 'bytes-array',
                len: 64,
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { createArrayU8Codec } from '@scale-codec/definition-runtime'

        import type { Codec } from '@scale-codec/definition-runtime'

        // Type: Hash

        type Hash = Uint8Array

        const Hash: Codec<Hash> = createArrayU8Codec('Hash', 64)

        // Exports

        export { Hash }"
    `)
})

it('Renders option', () => {
    expect(
        renderFactory({
            OptionHash: {
                t: 'option',
                some: 'Hash',
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { createOptionCodec, dynCodec } from '@scale-codec/definition-runtime'

        import type { EnumCodecAndFactory, Opaque, Option } from '@scale-codec/definition-runtime'

        // Dynamic codecs

        const __dyn_Hash = dynCodec(() => Hash)

        // Type: OptionHash

        interface OptionHash__actual extends Option<Hash> {}

        interface OptionHash extends Opaque<OptionHash__actual, OptionHash> {}

        const OptionHash: EnumCodecAndFactory<OptionHash> = createOptionCodec<OptionHash__actual, OptionHash>('OptionHash', __dyn_Hash)

        // Exports

        export { OptionHash }"
    `)
})

it('Renders empty struct as void alias', () => {
    expect(
        renderFactory({
            EmptyStruct: {
                t: 'struct',
                fields: [],
            },
        }),
    ).toMatchInlineSnapshot(`
        "// Type: EmptyStruct

        import { Void as EmptyStruct } from '@scale-codec/definition-runtime'

        // Exports

        export { EmptyStruct }"
    `)
})

it('Renders empty tuple as void alias', () => {
    expect(
        renderFactory({
            EmptyTuple: {
                t: 'tuple',
                items: [],
            },
        }),
    ).toMatchInlineSnapshot(`
        "// Type: EmptyTuple

        import { Void as EmptyTuple } from '@scale-codec/definition-runtime'

        // Exports

        export { EmptyTuple }"
    `)
})

it('Renders an alias for some inner type', () => {
    expect(
        renderFactory({
            StringAlias: {
                t: 'alias',
                ref: 'Str',
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { Str, dynCodec } from '@scale-codec/definition-runtime'

        import type { Codec } from '@scale-codec/definition-runtime'

        // Type: StringAlias

        interface StringAlias extends Str {}

        const StringAlias: Codec<StringAlias> = dynCodec(() => Str)

        // Exports

        export { StringAlias }"
    `)
})

it('Renders single tuple as alias in case when the related option is enabled', async () => {
    expect(
        renderFactory(
            {
                SingleTuple: { t: 'tuple', items: ['U128'] },
                MultiTuple: { t: 'tuple', items: ['U8', 'Bool'] },
            },

            {
                rollupSingleTuplesIntoAliases: true,
            },
        ),
    ).toMatchInlineSnapshot(`
        "import { Bool, U128, U8, createTupleCodec, dynCodec } from '@scale-codec/definition-runtime'

        import type { ArrayCodecAndFactory, Codec, Opaque } from '@scale-codec/definition-runtime'

        // Type: MultiTuple

        type MultiTuple__actual = [U8, Bool]

        interface MultiTuple extends Opaque<MultiTuple__actual, MultiTuple> {}

        const MultiTuple: ArrayCodecAndFactory<MultiTuple__actual, MultiTuple> = createTupleCodec<MultiTuple__actual, MultiTuple>('MultiTuple', [U8, Bool])

        // Type: SingleTuple

        interface SingleTuple extends U128 {}

        const SingleTuple: Codec<SingleTuple> = dynCodec(() => U128)

        // Exports

        export { MultiTuple, SingleTuple }"
    `)
})

it('Render import for the external type as expected', () => {
    expect(
        renderFactory({
            MyCustomExternal: {
                t: 'import',
                module: './module-with-externals',
            },
        }),
    ).toMatchInlineSnapshot(`
        "// Type: MyCustomExternal

        import { MyCustomExternal } from './module-with-externals'

        // Exports

        export { MyCustomExternal }"
    `)
})

it('Renders imports for the external type using the custom name if provided', () => {
    expect(
        renderFactory({
            ReExportMe: {
                t: 'import',
                module: 'some-package',
                nameInModule: 're_export_me',
            },
        }),
    ).toMatchInlineSnapshot(`
        "// Type: ReExportMe

        import { re_export_me as ReExportMe } from 'some-package'

        // Exports

        export { ReExportMe }"
    `)
})

it('Renders result', () => {
    expect(
        renderFactory({
            ResultI128Str: {
                t: 'result',
                ok: 'I128',
                err: 'Str',
            },
        }),
    ).toMatchInlineSnapshot(`
        "import { I128, Str, createResultCodec } from '@scale-codec/definition-runtime'

        import type { EnumCodecAndFactory, Opaque, Result } from '@scale-codec/definition-runtime'

        // Type: ResultI128Str

        interface ResultI128Str__actual extends Result<I128, Str> {}

        interface ResultI128Str extends Opaque<ResultI128Str__actual, ResultI128Str> {}

        const ResultI128Str: EnumCodecAndFactory<ResultI128Str> = createResultCodec<ResultI128Str__actual, ResultI128Str>('ResultI128Str', I128, Str)

        // Exports

        export { ResultI128Str }"
    `)
})

it('Respects custom `runtimeLib` param', () => {
    expect(
        renderFactory(
            {
                S: {
                    t: 'alias',
                    ref: 'Str',
                },
            },

            { runtimeLib: 'custom-runtime-lib' },
        ),
    ).toMatchInlineSnapshot(`
        "import { Str, dynCodec } from 'custom-runtime-lib'

        import type { Codec } from 'custom-runtime-lib'

        // Type: S

        interface S extends Str {}

        const S: Codec<S> = dynCodec(() => Str)

        // Exports

        export { S }"
    `)
})

it('Respects custom set of knowns types set', () => {
    expect(
        renderFactory(
            {
                FooAlias: {
                    t: 'alias',
                    ref: 'Foo',
                },

                // Should not be imported from runtime lib
                StrAlias: {
                    t: 'alias',
                    ref: 'Str',
                },
            },

            { runtimeTypes: new Set(['Foo']) },
        ),
    ).toMatchInlineSnapshot(`
        "import { Foo, dynCodec } from '@scale-codec/definition-runtime'

        import type { Codec } from '@scale-codec/definition-runtime'

        // Type: FooAlias

        interface FooAlias extends Foo {}

        const FooAlias: Codec<FooAlias> = dynCodec(() => Foo)

        // Type: StrAlias

        interface StrAlias extends Str {}

        const StrAlias: Codec<StrAlias> = dynCodec(() => Str)

        // Exports

        export { FooAlias, StrAlias }"
    `)
})

it('When type for void aliases is defined, it is used', () => {
    expect(
        renderFactory(
            {
                EmptyTuple: {
                    t: 'tuple',
                    items: [],
                },
            },

            { typeForVoidAliasing: 'SomeCustomVoid' },
        ),
    ).toMatchInlineSnapshot(`
        "// Type: EmptyTuple

        import { SomeCustomVoid as EmptyTuple } from '@scale-codec/definition-runtime'

        // Exports

        export { EmptyTuple }"
    `)
})
