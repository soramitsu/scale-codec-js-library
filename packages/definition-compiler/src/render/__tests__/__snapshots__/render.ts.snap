// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render import for the external type as expected 1`] = `
"/* eslint-disable */

// MyCustomExternal

import { MyCustomExternal_Decoded, MyCustomExternal_Encodable, MyCustomExternal_decode, MyCustomExternal_encode } from './module-with-externals'

export { MyCustomExternal_Decoded, MyCustomExternal_Encodable, MyCustomExternal_decode, MyCustomExternal_encode }
"
`;

exports[`Renders an alias for some inner type 1`] = `
"/* eslint-disable */

import { DecodeResult, str_Decoded, str_Encodable, str_decode, str_encode } from '@scale-codec/definition-runtime'

// StringAlias

export type StringAlias_Decoded = str_Decoded

export type StringAlias_Encodable = str_Encodable

export function StringAlias_decode(bytes: Uint8Array): DecodeResult<StringAlias_Decoded> {
    return str_decode(bytes)
}

export function StringAlias_encode(encodable: StringAlias_Encodable): Uint8Array {
    return str_encode(encodable)
}
"
`;

exports[`Renders array 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, decodeArray, encodeArray, makeEncoderAsIsRespectable } from '@scale-codec/definition-runtime'

// Array_Item_15

export type Array_Item_15_Decoded = Item_Decoded[]

export type Array_Item_15_Encodable = (Item_Encodable | EncodeAsIs)[]

const __hoisted_Array_Item_15_item_encode__ = makeEncoderAsIsRespectable(Item_encode)
const __hoisted_Array_Item_15_len__ = 15

export function Array_Item_15_decode(bytes: Uint8Array): DecodeResult<Array_Item_15_Decoded> {
    return decodeArray(bytes, Item_decode, __hoisted_Array_Item_15_len__)
}

export function Array_Item_15_encode(encodable: Array_Item_15_Encodable): Uint8Array {
    return encodeArray(encodable, __hoisted_Array_Item_15_item_encode__, __hoisted_Array_Item_15_len__)
}
"
`;

exports[`Renders bytes array 1`] = `
"/* eslint-disable */

import { DecodeResult, decodeUint8Array, encodeUint8Array } from '@scale-codec/definition-runtime'

// Hash

export type Hash_Decoded = Uint8Array

export type Hash_Encodable = Uint8Array

const __hoisted_Hash_len__ = 64

export function Hash_decode(bytes: Uint8Array): DecodeResult<Hash_Decoded> {
    return decodeUint8Array(bytes, __hoisted_Hash_len__)
}

export function Hash_encode(encodable: Hash_Encodable): Uint8Array {
    return encodeUint8Array(encodable, __hoisted_Hash_len__)
}
"
`;

exports[`Renders empty struct as void alias 1`] = `
"/* eslint-disable */

// EmptyStruct

import { Void_Decoded as EmptyStruct_Decoded, Void_Encodable as EmptyStruct_Encodable, Void_decode as EmptyStruct_decode, Void_encode as EmptyStruct_encode } from '@scale-codec/definition-runtime'

export { EmptyStruct_Decoded, EmptyStruct_Encodable, EmptyStruct_decode, EmptyStruct_encode }
"
`;

exports[`Renders empty tuple as void alias 1`] = `
"/* eslint-disable */

// EmptyTuple

import { Void_Decoded as EmptyTuple_Decoded, Void_Encodable as EmptyTuple_Encodable, Void_decode as EmptyTuple_decode, Void_encode as EmptyTuple_encode } from '@scale-codec/definition-runtime'

export { EmptyTuple_Decoded, EmptyTuple_Encodable, EmptyTuple_decode, EmptyTuple_encode }
"
`;

exports[`Renders enum 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, Enum, HelperEnumDiscriminantVariantPair, Valuable, decodeEnum, encodeEnum, helperEnumDecoders, helperEnumEncoders, str_Decoded, str_Encodable, str_decode, str_encode } from '@scale-codec/definition-runtime'

// Message

export type Message_Decoded = Enum<{
    Quit: null
    Greeting: Valuable<str_Decoded>
}>

export type Message_Encodable = Enum<{
    Quit: null
    Greeting: Valuable<str_Encodable | EncodeAsIs>
}>

const __hoisted_Message_pairs__: HelperEnumDiscriminantVariantPair[] = [
    [0, 'Quit'],
    [1, 'Greeting']
]
const __hoisted_Message_decoders__ = helperEnumDecoders(__hoisted_Message_pairs__, {
    1: str_decode
})
const __hoisted_Message_encoders__ = helperEnumEncoders(__hoisted_Message_pairs__, {
    Greeting: str_encode
})

export function Message_decode(bytes: Uint8Array): DecodeResult<Message_Decoded> {
    return decodeEnum(bytes, __hoisted_Message_decoders__)
}

export function Message_encode(encodable: Message_Encodable): Uint8Array {
    return encodeEnum(encodable, __hoisted_Message_encoders__)
}
"
`;

exports[`Renders imports for the external type using the custom name if provided 1`] = `
"/* eslint-disable */

// ReExportMe

import { re_export_me_Decoded as ReExportMe_Decoded, re_export_me_Encodable as ReExportMe_Encodable, re_export_me_decode as ReExportMe_decode, re_export_me_encode as ReExportMe_encode } from 'some-package'

export { ReExportMe_Decoded, ReExportMe_Encodable, ReExportMe_decode, ReExportMe_encode }
"
`;

exports[`Renders map 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, decodeMap, encodeMap, helperMapEncoders, str_Decoded, str_Encodable, str_decode, str_encode } from '@scale-codec/definition-runtime'

// Map_str_Message

export type Map_str_Message_Decoded = Map<str_Decoded, Message_Decoded>

export type Map_str_Message_Encodable = Map<str_Encodable | EncodeAsIs, Message_Encodable | EncodeAsIs>

const [__hoisted_Map_str_Message_encode_key__, __hoisted_Map_str_Message_encode_value__] = helperMapEncoders(str_encode, Message_encode)

export function Map_str_Message_decode(bytes: Uint8Array): DecodeResult<Map_str_Message_Decoded> {
    return decodeMap(bytes, str_decode, Message_decode)
}

export function Map_str_Message_encode(encodable: Map_str_Message_Encodable): Uint8Array {
    return encodeMap(encodable, __hoisted_Map_str_Message_encode_key__, __hoisted_Map_str_Message_encode_value__)
}
"
`;

exports[`Renders option 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, HelperEnumDiscriminantVariantPair, Option, decodeEnum, encodeEnum, helperEnumDecoders, helperEnumEncoders } from '@scale-codec/definition-runtime'

// OptionHash

export type OptionHash_Decoded = Option<Hash_Decoded>

export type OptionHash_Encodable = Option<Hash_Encodable | EncodeAsIs>

const __hoisted_OptionHash_pairs__: HelperEnumDiscriminantVariantPair[] = [
    [0, 'None'],
    [1, 'Some']
]
const __hoisted_OptionHash_decoders__ = helperEnumDecoders(__hoisted_OptionHash_pairs__, {
    1: Hash_decode
})
const __hoisted_OptionHash_encoders__ = helperEnumEncoders(__hoisted_OptionHash_pairs__, {
    Some: Hash_encode
})

export function OptionHash_decode(bytes: Uint8Array): DecodeResult<OptionHash_Decoded> {
    return decodeEnum(bytes, __hoisted_OptionHash_decoders__)
}

export function OptionHash_encode(encodable: OptionHash_Encodable): Uint8Array {
    return encodeEnum(encodable, __hoisted_OptionHash_encoders__)
}
"
`;

exports[`Renders set 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, decodeSet, encodeSet, makeEncoderAsIsRespectable } from '@scale-codec/definition-runtime'

// Set_Message

export type Set_Message_Decoded = Set<Message_Decoded>

export type Set_Message_Encodable = Set<Message_Encodable | EncodeAsIs>

const Set_Message_item_encode = makeEncoderAsIsRespectable(Message_encode)

export function Set_Message_decode(bytes: Uint8Array): DecodeResult<Set_Message_Decoded> {
    return decodeSet(bytes, Message_decode)
}

export function Set_Message_encode(encodable: Set_Message_Encodable): Uint8Array {
    return encodeSet(encodable, Set_Message_item_encode)
}
"
`;

exports[`Renders single tuple as alias in case when the related option is enabled 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, bool_Decoded, bool_Encodable, bool_decode, bool_encode, decodeTuple, encodeTuple, helperTupleEncoders, u128_Decoded, u128_Encodable, u128_decode, u128_encode, u8_Decoded, u8_Encodable, u8_decode, u8_encode } from '@scale-codec/definition-runtime'

// MultiTuple

export type MultiTuple_Decoded = [u8_Decoded, bool_Decoded]

export type MultiTuple_Encodable = [u8_Encodable | EncodeAsIs, bool_Encodable | EncodeAsIs]

const __hoisted_MultiTuple_decoders__ = [u8_decode, bool_decode]
const __hoisted_MultiTuple_encoders__ = helperTupleEncoders<MultiTuple_Decoded>([u8_encode, bool_encode])

export function MultiTuple_decode(bytes: Uint8Array): DecodeResult<MultiTuple_Decoded> {
    return decodeTuple(bytes, __hoisted_MultiTuple_decoders__ as any)
}

export function MultiTuple_encode(encodable: MultiTuple_Encodable): Uint8Array {
    return encodeTuple(encodable, __hoisted_MultiTuple_encoders__ as any)
}

// SingleTuple

export type SingleTuple_Decoded = u128_Decoded

export type SingleTuple_Encodable = u128_Encodable

export function SingleTuple_decode(bytes: Uint8Array): DecodeResult<SingleTuple_Decoded> {
    return u128_decode(bytes)
}

export function SingleTuple_encode(encodable: SingleTuple_Encodable): Uint8Array {
    return u128_encode(encodable)
}
"
`;

exports[`Renders struct + tuple 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, bool_Decoded, bool_Encodable, bool_decode, bool_encode, decodeStruct, decodeTuple, encodeStruct, encodeTuple, helperStructEncoders, helperTupleEncoders, i128_Decoded, i128_Encodable, i128_decode, i128_encode, u8_Decoded, u8_Encodable, u8_decode, u8_encode } from '@scale-codec/definition-runtime'

// Structural

export type Structural_Decoded = { nums: Tuple_u8_i128_Decoded, mur: bool_Decoded }

export type Structural_Encodable = { nums: Tuple_u8_i128_Encodable | EncodeAsIs, mur: bool_Encodable | EncodeAsIs }

const __hoisted_Structural_order__: (keyof Structural_Decoded)[] = ['nums', 'mur']
const __hoisted_Structural_decoders__ = { nums: Tuple_u8_i128_decode, mur: bool_decode }
const __hoisted_Structural_encoders__ = helperStructEncoders({ nums: Tuple_u8_i128_encode, mur: bool_encode })

export function Structural_decode(bytes: Uint8Array): DecodeResult<Structural_Decoded> {
    return decodeStruct(bytes, __hoisted_Structural_decoders__, __hoisted_Structural_order__)
}

export function Structural_encode(encodable: Structural_Encodable): Uint8Array {
    return encodeStruct(encodable, __hoisted_Structural_encoders__, __hoisted_Structural_order__)
}

// Tuple_u8_i128

export type Tuple_u8_i128_Decoded = [u8_Decoded, i128_Decoded]

export type Tuple_u8_i128_Encodable = [u8_Encodable | EncodeAsIs, i128_Encodable | EncodeAsIs]

const __hoisted_Tuple_u8_i128_decoders__ = [u8_decode, i128_decode]
const __hoisted_Tuple_u8_i128_encoders__ = helperTupleEncoders<Tuple_u8_i128_Decoded>([u8_encode, i128_encode])

export function Tuple_u8_i128_decode(bytes: Uint8Array): DecodeResult<Tuple_u8_i128_Decoded> {
    return decodeTuple(bytes, __hoisted_Tuple_u8_i128_decoders__ as any)
}

export function Tuple_u8_i128_encode(encodable: Tuple_u8_i128_Encodable): Uint8Array {
    return encodeTuple(encodable, __hoisted_Tuple_u8_i128_encoders__ as any)
}
"
`;

exports[`Renders vec 1`] = `
"/* eslint-disable */

import { DecodeResult, EncodeAsIs, decodeVec, encodeVec, makeEncoderAsIsRespectable, str_Decoded, str_Encodable, str_decode, str_encode } from '@scale-codec/definition-runtime'

// Vec_str

export type Vec_str_Decoded = str_Decoded[]

export type Vec_str_Encodable = (str_Encodable | EncodeAsIs)[]

const __hoisted_Vec_str_item_encode__ = makeEncoderAsIsRespectable(str_encode)

export function Vec_str_decode(bytes: Uint8Array): DecodeResult<Vec_str_Decoded> {
    return decodeVec(bytes, str_decode)
}

export function Vec_str_encode(encodable: Vec_str_Encodable): Uint8Array {
    return encodeVec(encodable, __hoisted_Vec_str_item_encode__)
}
"
`;
